### ISSUE TITLE
Possible typo in Projector/ProjectorAM2Order.cpp

### ISSUE THREAD

**Issue by Status-Mirror** [SmileiPIC/Smilei#642]
Hi all,

I've been going through the source-code trying to understand the cylindrical-PIC algorithm, and there's a line in `Projector/ProjectorAM2Order.cpp` which looks wrong to me. On lines 140 and 834, we find:

```
 double r_bar = ((jpo + j_domain_begin_)*dr + deltaold[1*nparts] + rp) * 0.5; // r at t = t0 - dt/2
```

From the comment, I would guess this is trying to average the radial particle position before the particle push, and the radial position after. My understanding of these terms are as follows:

- `(jpo + j_domain_begin_)*dr`: Radial position of closest cell-edge to the particle position before the most recent push.
- `deltaold[1*nparts]`: Radial distance from the particle position to the previous closest cell-edge, **divided by dr**.
- `rp`: Radial position of the particle after the push.

I think this line is attempting to work out $\bar{r}=(r_{old} + r_{new})/2$, but I don't believe this is doing it. Should the line instead read:

```
 double r_bar = ((jpo + j_domain_begin_ + deltaold[1*nparts])*dr + rp) * 0.5; // r at t = t0 - dt/2
```

I'm still working through the source-code and my understanding is incomplete. There could be a good reason why the line is as it is, but I think it's likely there's a typo/bug here.

Cheers,
Stuart

**Comment by Z10Frank** [comment]
Hello,
thank you very much! 
We'll check those lines as soon as possible.

**Comment by beck-llr** [comment]
Hi Stuart

You are correct on both points.
 I have tried a few tests with the corrections you propose and saw very little difference in the results. The most important difference I see is the noise of `Jt` mode 0 being significantly higher. It was probably underestimated before.

Nevertheless, these corrections should be introduced. I'm just going to run a couple more tests and you should see them appear soon in the development branch. 

I'd like to thank you very much for the very careful reading of the code and for reporting your findings to us in a crystal clear post. Congratulations to you and I really look forward to reading your next suggestions to improve the code quality :-)
Also if you have questions about some part of the code, do not hesitate to contact us on the chat.

Cheers

P.S: I'll close the issue when the corrections are available.

**Comment by Status-Mirror** [comment]
Hi @beck-llr 

Thanks for the reply, it's good to know that my understanding of the code is correct! I have now finished my review through the code, and I think I understand all the equations which are relevant to the core cylindrical-PIC algorithm. During this review I found a few more points which could also be potential bugs and typos, and I'm happy to share my findings with you.

Initially I tried typing this out here, but I hit some kind of limit on the GitHub markdown service. The response became so long that equations stopped formatting correctly! As a workaround, I'll provide a link to the report on Overleaf, as I can't seem to attach the PDF here.

Smilei bug report: https://www.overleaf.com/read/hptynqbzzrpj

Thanks for all the help,
Stuart

**Comment by Status-Mirror** [comment]
On point 2, I agree - I've discovered a contradiction in my argument. The macro-particles have no "shape" in the $\theta$ direction, as the interpolator just uses $\exp(im\theta)$ without averaging over nearby $\theta$. Hence, when calculating the current density contribution from a single macro-particle, I took $J(x,r,\theta) = J(x,r)\delta(\theta-\theta_p)$ where $\theta_p$ is the particle $\theta$. Solving this integral did not introduce a cancelling $2\pi$ factor. The error in my approach was that if I take a macro-particle shape with a missing dimension, then the _macroparticle_ number density increases by a factor of $2\pi$, which provides the cancellation factor.

I see your approach to point 3 is provided in your documentation, but I would argue there is an inconsistency between the treatment of the fields and the current densities (see my bug 6 discussion). Your conditions are correct for the goal: "If you want $\hat{B}_r^m$ to be zero", but I don't think you should force that. If the macro-particle shape can deposit current below $r=0$ in a physical way, then why can't the fields below $r=0$ influence acceleration on the macro-particle?

Following current density logic, I think the interpolator should give:

$$B_r^m = \left( 0.75 \hat{B}_r^m[2] + 0.125 \hat{B}_r^m[3] + 0.125 \hat{B}_r^m[1]\right) \exp(im\theta)$$

where, from the geometry of the physical space, $\hat{B}_r^m[1] \exp(im\theta) = -\hat{B}_r^m[3] \exp(im(\theta+\pi))$. This would suggest: 

$$B_r^m = \left( 0.75 \hat{B}_r^m[2] + 0.125 \left( \hat{B}_r^m[3] - (-1)^m \hat{B}_r^m[3]\right)\right) \exp(im\theta)$$

Using this interpolation method would make everything consistent, but maybe there's a reason you've chosen not to use it? I see that for a particle with exactly $r=0$, then $\theta$ becomes undefined. Perhaps you've chosen the non-physical method to prevent errors in evaluating $\exp(im\theta)$ when $r=0$?

Ultimately I think point 3 is an implementation choice, and I don't think the differences in our approach would radically change the output of the simulation.

Cheers,
Stuart

**Comment by beck-llr** [comment]
For the point 4.1 you are correct, the minus sign must be replaced by a plus. Congrats for spotting this one out !

**Comment by mccoys** [comment]
@beck-llr anything left to do in this issue?

**Comment by beck-llr** [comment]
Yes I'm not done yet. Unfortunately this takes time and I haven't had time to prepare a proper answer to the interesting points raised here.

**Comment by beck-llr** [comment]
@Status-Mirror About point 4.2: missing currents in the Buneman boundary conditions.

First please note that the boundary conditions on particles are applied before current deposition. Therefore a particle crossing r=rmax would be removed or reflected before it has a chance to deposit currents as far as you suggest.

Moreover the Buneman boundary conditions are not derived with the objective to solve locally Maxwell's equations but to analytically try to force minimum reflections of outgoing waves back into the domain.  In order to do so you have to assume some conditions on the state of the plasma "outside" the simulation domain and that assumption is that there is no current. Of course if there is a strong current at the boundary this assumption becomes false and this is a limitation of this method.

**Comment by beck-llr** [comment]
@Status-Mirror About point 4.3: you are correct the `-1` here is kind of arbitrary and not really justified.
 The question of boundary conditions in the corners is a very difficult one since several boundary conditions could possibly apply on top of each other and therefore overwriting each other. One has to choose which boundary is the most relevant. It has been decided that longitudinal boundaries are. And that is why the loop for $B_t$ in the very same function starts at 1 and finished at $n_p$ (which is at 1 cell before the end since $B_t$ is dual along x). So that the extreme points of $B_t$ are defined by the longitudinal boundaries and not the radial ones.
A similar logic does not apply for $B_l$ since it is not affected by longitudinal boundaries hence the loop starts at 0 and should finish at $n_p$ as you noticed.

**Comment by beck-llr** [comment]
@Status-Mirror About 5.1 you are absolutely correct. I've run some basic tests and this correction has some impact on the Bl component of the reflected laser on the xmax boundary. Again, congratulation for spotting it and thanks a lot for reporting it.

5.2 is obviously an ugly typo :-)

**Comment by charlesprouveur** [comment]
@beck-llr is there anything left to do in this issue or we can close it?

**Comment by beck-llr** [comment]
Somme additional comments have been addressed but I haven't had time to answer here yet. Let's keep it open.
